designs/documents.md

```diff
@@ -0,0 +1,424 @@
+
+By default, the document's schema is a generic schema from the Smithy prelude is
+based on the underlying type. A string shape takes on the base `String` for
+example. Lists and maps by default use a schema with the shape type `DOCUMENT`.
+Since it is not possible to determin whether a given `dict` is a map, structure,
+or union, they are treated as maps by default. See
+[structures and unions](#structures-and-unions) for more information on how
+typed structure and union `Document`s may be created.
+
+It is also possible to pass in a schema when constructing a document by using
+the `schema` keyword-only argument.
+
+### Accessors
+
+The `Document` type has accessors for each Python data type that is generated
+from Smithy. This means, for example, that there is an `as_integer`, but not an
+`as_short` or `as_big_decimal`. If it is necessary to distinguish between the
+different integer or float sizes, the `shape_type` can be used.
+
+When any of the accessor methods are called, the underlying data is asserted to
+be of the requested type before it is returned, raising an exception if it is
+not. In some cases these types are coerced where they're compatible. For
+example, attempting to access a `Decimal` from a document that contains a
+`float` will return a `Decimal` representation of that float.
+
+The following example shows how accessors may be used:
+
+```python
+# With no or limited prior knowledge, a match can be used to handle different
+# possible types.
+match document.shape_type:
+    case ShapeType.STRING:
+        print(document.as_string())
+    case ShapeType.BLOB:
+        print(document.as_blob().decode("utf-8"))
+    case _:
+        pass
+
+# With concrete prior knowledge, the accessor can be called directly.
+# If the underlying type doesn't match, an exception will be raised.
+print(f"Total: {document.as_float()}")
+```
+
+Note that the values of lists and maps are `Document`s themselves.
+
+#### DocumentValue
+
+In addition to accessing Smithy data types, `Document`s will have a method to
+return the document as a plain Python object without any context. These
+`DocumentValue`s are similar to a JSON value, except that they include
+`datetime.datetime`, `bytes`, and `Decimal` as possible types. `Doucment`s may
+also be construced by providing a `DocumentValue`. Other implementations of the
+`Document` type may not include this functionality, but for Python it is a huge
+ease-of-use improvement for those that don't need the extra saftey and
+functionality of the full `Document` type.
+
+Currently the base `Document` implementation lazily converts between the two
+representations and caches the results. This is done for the sake of speed, but
+has an unfortunate space cost. This behavior is not part of the API, however,
+and MAY be changed without backwards compatibility implications.
+
+```python
+>>> document = Document({"foo": "bar"})
+>>> document.shape_type
+ShapeType.DOCUMENT
+>>> document.as_value()
```

Comment:
What is the distinction here between a "value" and a python dictionary?

Reply:
It's not always a dictionary. Like JSON, there is nothing requiring a document to be rooted with a mapping or list. `True` is a valid `DocumentValue` but not a valid `dict` for example.